src/Time/Actions/LimitTimeStepToExpirationTimes.hpp

```diff
@@ -0,0 +1,175 @@
+///
+/// Uses:
+/// - DataBox:
+///   - Tags::Time
+///   - Tags::TimeStep
+///   - Tags::TimeStepId
+///   - Tags::TimeStepper<>
+///   - domain::Tags::FunctionsOfTime
+///
+/// DataBox changes:
+/// - Adds: nothing
+/// - Removes: nothing
+/// - Modifies:
+///   - Tags::TimeStep
+struct LimitTimeStepToExpirationTimes {
+  template <typename DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static Parallel::iterable_action_return_t apply(
+      db::DataBox<DbTags>& box, tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      Parallel::GlobalCache<Metavariables>& cache,
+      const ArrayIndex& /*array_index*/, ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) {  // NOLINT const
+    // First, check whether the time stepper uses substeps
+    const TimeStepper& time_stepper = db::get<Tags::TimeStepper<>>(box);
+    if (time_stepper.number_of_substeps() > 0) {
+      // Next, check if the current time step is at the beginning of a full step
+      const auto& time_step_id = db::get<Tags::TimeStepId>(box);
+      if (time_step_id.substep() == 0) {
+        // Get the functions of time and, if not empty, loop over them,
+        // finding the minimum expiration time
+        const auto& functions_of_time =
+            get<domain::Tags::FunctionsOfTime>(cache);
+        if (not functions_of_time.empty()) {
+          double min_expiration_time{std::numeric_limits<double>::max()};
+          for (const auto& [name, f_of_t] : functions_of_time) {
+            if (f_of_t->time_bounds()[1] < min_expiration_time) {
+              min_expiration_time = f_of_t->time_bounds()[1];
+            }
+          }
+
+          ASSERT(db::get<Tags::Time>(box) <= min_expiration_time,
+                 "Time (" << std::setprecision(20) << db::get<Tags::Time>(box)
+                          << ") is greater than expiration time ("
+                          << min_expiration_time << ") by an amount of "
+                          << db::get<Tags::Time>(box) - min_expiration_time);
+          if (db::get<Tags::Time>(box) == min_expiration_time) {
+            return {Parallel::AlgorithmExecution::Continue, std::nullopt};
+          }
+
+          const auto& initial_time_step{db::get<Tags::TimeStep>(box)};
+          const double step_end =
+              (time_step_id.step_time() + initial_time_step).value();
+          if (time_step_id.step_time().value() +
+                  2.0 * initial_time_step.value() <=
+              min_expiration_time) {
+            // The expiration times are far in the future (more than
+            // two steps).  No need to adjust anything.
+            return {Parallel::AlgorithmExecution::Continue, std::nullopt};
+          }
+
+          const double start{initial_time_step.slab().start().value()};
+          Slab new_slab{};
+          if (step_end + slab_rounding_error(time_step_id.step_time()) >=
```

Comment:
I know I wrote this, but I've changed my mind and don't like this increasing the step size, even by a small amount.  Please drop the `slab_rounding_error` term and update the comment below.

The slab-resizing code doesn't handle roundoff very well at the moment, but I'm doing a bunch of cleanup there at the moment and hope I'll get around to fixing that.  If I do, increasing the size here could confuse other code.